### PR TITLE
Feature/issues/402

### PR DESCRIPTION
**Pull request by Mizzick**
#402

### CODE REVIEW

File: Extension/lib/filter/filters.js [hunk 1]
```diff
@@ -246,7 +246,7 @@ RequestFilter.prototype = {
             return cacheItem[0];
         }
 
-        var rule = this._checkWhiteList(requestUrl, refHost, requestType, thirdParty);
+        var rule = this._checkWhiteList(requestUrl, referrer, requestType, thirdParty);
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: Why referrer? We have already extracted `refHost`

File: Extension/lib/filter/filters.js [hunk 2]
```diff
@@ -255,12 +255,13 @@ RequestFilter.prototype = {
     /**
      * Searches for the filter rule for the specified request.
      *
-     * @param requestUrl    Request URL
-     * @param referrer      Referrer
-     * @param requestType   Request content type (one of UrlFilterRule.contentTypes)
+     * @param requestUrl            Request URL
+     * @param referrer              Referrer
+     * @param requestType           Request content type (one of UrlFilterRule.contentTypes)
+     * @param documentWhitelistRule Document-level whitelist rule
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: Mark is as optional

File: Extension/lib/filter/filters.js [hunk 3]
```diff
@@ -272,7 +273,7 @@ RequestFilter.prototype = {
             return cacheItem[0];
         }
 
-        var rule = this._innerFilterHttpRequest(requestUrl, referrer, refHost, requestType, thirdParty);
+        var rule = this._innerFilterHttpRequest(requestUrl, referrer, refHost, requestType, thirdParty, documentWhitelistRule);
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: Rename this method into something making sense

File: Extension/lib/filter/filters.js [hunk 4]
```diff
@@ -365,33 +366,53 @@ RequestFilter.prototype = {
      * @param refHost       Referrer host
      * @param requestType   Request content type (one of UrlFilterRule.contentTypes)
      * @param thirdParty    Is request third-party or not
+     * @param documentWhitelistRule Document-level whitelist rule
      * @returns Filter rule found or null
      * @private
      */
-    _innerFilterHttpRequest: function (requestUrl, referrer, refHost, requestType, thirdParty) {
+    _innerFilterHttpRequest: function (requestUrl, referrer, refHost, requestType, thirdParty, urlWhiteListRule) {
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: Name in comment is different

File: Extension/lib/filter/filters.js [hunk 5]
```diff
@@ -365,33 +366,53 @@ RequestFilter.prototype = {
...
 
-        var urlWhiteRule = this._checkWhiteList(requestUrl, refHost, requestType, thirdParty);
-        if (urlWhiteRule !== null) {
-            Log.debug("White list rule found {0} for url: {1} referrer: {2}, requestType: {3}", urlWhiteRule.ruleText, requestUrl, refHost, requestType);
-            return urlWhiteRule;
+        // STEP 1: Looking for exception rule, which could be applied to the whole document, request was originated from
+
+        // Searching white list for a rule for Referrer and checking it's UrlBlock attribute
+        // If UrlBlock is set - than we should not use UrlBlockingFilter against this request.
+        var referrerWhiteListRule = this._checkWhiteList(referrer, referrer, "DOCUMENT", thirdParty);
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: What is it?

Comment (ameshkov) [inline Extension/lib/filter/filters.js]: I see you've missed the whole point with NOT USING REFERRER

File: Extension/lib/filter/request-blocking.js [hunk 6]
```diff
@@ -32,207 +32,270 @@ var WebRequestService = exports.WebRequestService = function (framesMap, antiBan
...
-    if (safariContentBlockerEnabled && collapseAllElements) {
-        // For Safari 9+ we will load all selectors when browser is just started
-        // as at that moment content blocker may not been initialized
-        return true;
-    }
-
-    // In other cases we should load all selectors every time
-    return !safariContentBlockerEnabled;
-};
+        var jsInjectRule = this.antiBannerService.getRequestFilter().findWhiteListRule(documentUrl, documentUrl, "JSINJECT");
```
Comment (ameshkov) [inline Extension/lib/filter/request-blocking.js]: You have got `whitelistRule` above, what's going on, why are you doing it?

File: Extension/lib/filter/rules/url-filter-rule.js [hunk 7]
```diff
@@ -172,6 +181,35 @@ UrlFilterRule.prototype.isFiltered = function (requestUrl, thirdParty, requestCo
...
+    var contentTypeMask = UrlFilterRule.contentTypes[contentType];
+    return this.checkContentTypeMask(contentTypeMask);
+};
+
+/**
+ * Checks if specified content type is suitable.
+ *
+ * @param contentType Request content type mask (UrlFilterRule.contentTypes)
+ */
+UrlFilterRule.prototype.checkContentTypeMask = function (contentTypeMask) {
```
Comment (ameshkov) [inline Extension/lib/filter/rules/url-filter-rule.js]: This method is used in checkContentType only.

You could simply merge this code there.

File: Extension/lib/filter/rules/url-filter-rule.js [hunk 8]
```diff
@@ -283,12 +331,13 @@ UrlFilterRule.contentTypes = {
...
-    POPUP: 1 << 23,      //check block popups
-    GENERICHIDE: 1 << 24, //CssFilter generic rules cannot be applied to page
-    GENERICBLOCK: 1 << 25 //UrlFilter generic rules cannot be applied to page
+    ELEMHIDE: 1 << 20,      //CssFilter cannot be applied to page
+    URLBLOCK: 1 << 21,      //This attribute is only for exception rules. If true - do not use urlblocking rules for urls where referrer satisfies this rule.
+    JSINJECT: 1 << 22,      //Does not inject javascript rules to page
+    POPUP: 1 << 23,         //check block popups
+    GENERICHIDE: 1 << 24,   //CssFilter generic rules cannot be applied to page
+    GENERICBLOCK: 1 << 25,  //UrlFilter generic rules cannot be applied to page
+    IMPORTANT: 1 << 26      //Important rules cannot be applied to page
```
Comment (ameshkov) [inline Extension/lib/filter/rules/url-filter-rule.js]: Why do we need it here?

File: Extension/lib/filter/rules/url-filter-rule.js [hunk 9]
```diff
@@ -311,6 +360,7 @@ UrlFilterRule.ignoreOptions = {
 };
 
 UrlFilterRule.contentTypes.DOCUMENT = UrlFilterRule.contentTypes.ELEMHIDE | UrlFilterRule.contentTypes.URLBLOCK | UrlFilterRule.contentTypes.JSINJECT;
+UrlFilterRule.contentTypes.DOCUMENT_LEVEL = UrlFilterRule.contentTypes.DOCUMENT | UrlFilterRule.contentTypes.GENERICHIDE | UrlFilterRule.contentTypes.GENERICBLOCK;
```
Comment (ameshkov) [inline Extension/lib/filter/rules/url-filter-rule.js]: I suggest renaming it to `DOCUMENT_LEVEL_EXCEPTIONS`

It should also contain:
`URLBLOCK`, `ELEMHIDE`, `JSINJECT`

File: Extension/lib/filter/filters.js [hunk 10]
```diff
@@ -258,7 +258,7 @@ RequestFilter.prototype = {
      * @param requestUrl            Request URL
      * @param referrer              Referrer
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: I'd prefer to name it `documentUrl` as it is referrer header value.

File: Extension/lib/filter/filters.js [hunk 11]
```diff
@@ -326,17 +326,17 @@ RequestFilter.prototype = {
      * Checks if exception rule is present for the URL/Referrer pair
      *
      * @param requestUrl    Request URL
-     * @param referrer      Referrer
+     * @param referrerHost  Referrer host
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: The same, `documentHost` seems to better describe what it is

File: Extension/lib/filter/filters.js [hunk 12]
```diff
@@ -359,30 +359,30 @@ RequestFilter.prototype = {
     },
 
     /**
-     * Filters HTTP request.
+     * Searches the rule for request.
      *
      * @param requestUrl    Request URL
      * @param referrer      Referrer
      * @param refHost       Referrer host
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: `documentHost`

File: Extension/lib/filter/filters.js [hunk 13]
```diff
@@ -359,30 +359,30 @@ RequestFilter.prototype = {
...
         Log.debug("Filtering http request for url: {0}, referrer: {1}, requestType: {2}", requestUrl, refHost, requestType);
 
-        // STEP 1: Looking for exception rule, which could be applied to the whole document, request was originated from
+        // STEP 1: Looking for exception rule, which could be applied to the current request
 
-        // Searching white list for a rule for Referrer and checking it's UrlBlock attribute
-        // If UrlBlock is set - than we should not use UrlBlockingFilter against this request.
-        var referrerWhiteListRule = this._checkWhiteList(referrer, referrer, "DOCUMENT", thirdParty);
+        // Checks white list for a rule for this RequestUrl. If something is found - returning it.
+        var urlWhiteListRule = this._checkWhiteList(requestUrl, refHost, requestType, thirdParty);
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: Yep, that's right

File: Extension/lib/filter/filters.js [hunk 14]
```diff
@@ -400,12 +400,12 @@ RequestFilter.prototype = {
         }
 
         if (!genericRulesAllowed || !urlRulesAllowed) {
-            Log.debug("White list rule {0} found for referrer: {1}", referrerWhiteListRule.ruleText, referrer);
+            Log.debug("White list rule {0} found for referrer: {1}", documentWhiteListRule.ruleText, referrer);
```
Comment (ameshkov) [inline Extension/lib/filter/filters.js]: Not for referrer, for `document`

File: Extension/lib/filter/request-blocking.js [hunk 15]
```diff
@@ -71,19 +71,20 @@ WebRequestService.prototype = (function () {
         };
 
         var whitelistRule = this.framesMap.getFrameWhiteListRule(tab);
```
Comment (ameshkov) [inline Extension/lib/filter/request-blocking.js]: Just use the `whitelistRule` and check it for `generichide`, `elemhide` and `jsinject`.
You shouldn't look for another rules anymore.

File: Extension/lib/filter/rules/url-filter-lookup-table.js [hunk 16]
```diff
@@ -68,18 +68,17 @@ UrlFilterRuleLookupTable.prototype = {
      * Returns filtering rule if request is filtered or NULL if nothing found
      *
      * @param url                 Url to check
-     * @param referrer            Request referrer
+     * @param referrerHost        Request referrer host
```
Comment (ameshkov) [inline Extension/lib/filter/rules/url-filter-lookup-table.js]: `documentHost`

File: Extension/lib/filter/rules/url-filter.js [hunk 17]
```diff
@@ -82,18 +82,18 @@ UrlFilter.prototype = {
 	 * Searches for first rule matching specified request
 	 *
 	 * @param url           Request url
-	 * @param referrer      Referrer host
+	 * @param referrerHost  Referrer host
```
Comment (ameshkov) [inline Extension/lib/filter/rules/url-filter.js]: documentHost

**Comment by atropnikov** [comment]
In previous version was supposed that frameWhiteListRule in frame data had DOCUMENT modifier. Now it is wrong. Find usages of frameWhiteListRule and fix problems.

File: Extension/lib/filter/request-blocking.js [hunk 18]
```diff
@@ -32,207 +32,271 @@ var WebRequestService = exports.WebRequestService = function (framesMap, antiBan
...
-
-    var genericHideRule = genericHide || this.antiBannerService.getRequestFilter().findWhiteListRule(documentUrl, documentUrl, "GENERICHIDE");
-    var elemHideRule = this.antiBannerService.getRequestFilter().findWhiteListRule(documentUrl, documentUrl, "ELEMHIDE");
-    if (!elemHideRule) {
-        if (this.shouldLoadAllSelectors(result.collapseAllElements)) {
-            result.selectors = this.antiBannerService.getRequestFilter().getSelectorsForUrl(documentUrl, genericHideRule);
-        } else {
-            result.selectors = this.antiBannerService.getRequestFilter().getInjectedSelectorsForUrl(documentUrl, genericHideRule);
+        var whitelistRule = this.framesMap.getFrameWhiteListRule(tab);
+        var genericHideFlag = (whitelistRule && whitelistRule.checkContentType("GENERICHIDE"));
```
Comment (atropnikov) [inline Extension/lib/filter/request-blocking.js]:
```
var genericHideFlag = genericHide || (whitelistRule && whitelistRule.checkContentType("GENERICHIDE"));
```

File: Extension/lib/filter/rules/url-filter-lookup-table.js [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+        return null;
+    },
+
+    /**
+     * Checks if rule filters request
+     *
+     * @param rule                Rule
+     * @param referrerHost        Referrer host
+     * @param url                 Request url
+     * @param urlLowerCase        Request url in lower case
```
Comment (atropnikov) [inline Extension/lib/filter/rules/url-filter-lookup-table.js]: Unused parameter

File: Extension/lib/filter/rules/url-filter.js [hunk 20]
```diff
@@ -64,73 +63,46 @@ UrlFilter.prototype = {
...
 	 *
 	 * @param url           Request url
-	 * @param refHost       Referrer host
+	 * @param documentHost  Document host
 	 * @param requestType   Request content type (UrlFilterRule.contentTypes)
 	 * @param thirdParty    true if request is third-party
 	 * @param skipGenericRules    skip generic rules
 	 * @return matching rule or null if no match found
 	 */
-	isFiltered: function (url, refHost, requestType, thirdParty, skipGenericRules) {
```
Comment (atropnikov) [inline Extension/lib/filter/rules/url-filter.js]: Double negation is confusing. Use **genericRulesAllowed**.

Comment (Mizzick) [inline Extension/lib/filter/rules/url-filter.js]: this is done to use it with undeclared parameters in some cases, like tests etc.

File: Extension/lib/filter/rules/url-filter-rule.js [hunk 21]
```diff
@@ -172,6 +181,26 @@ UrlFilterRule.prototype.isFiltered = function (requestUrl, thirdParty, requestCo
 };
 
 /**
+ * Checks if specified content type is suitable.
+ *
+ * @param contentType Request content type (UrlFilterRule.contentTypes)
+ */
+UrlFilterRule.prototype.checkContentType = function (contentType) {
```
Comment (atropnikov) [inline Extension/lib/filter/rules/url-filter-rule.js]: This function doesn't work properly. Please, add tests.
```
var rule = new UrlFilterRule("@@||example.com$elemhide");
assert.ok(rule.checkContentType("ELEMHIDE"));
assert.notOk(rule.checkContentType("DOCUMENT"));
```

**Pull request merged**